lib/airbrake-ruby/promise.rb

```diff
@@ -0,0 +1,99 @@
+module Airbrake
+  ##
+  # Represents a simplified promise object (similar to promises found in
+  # JavaScript), which allows chaining callbacks that are executed when the
+  # promise is either resolved or rejected.
+  #
+  # @see https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise
+  # @see https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/promise.rb
+  # @since v1.7.0
+  class Promise
+    attr_reader :value, :reason
+
+    def initialize
+      @on_resolved = []
+      @on_rejected = []
+      @value = nil
+      @reason = nil
+      @mutex = Mutex.new
+    end
+
+    ##
+    # Attaches a callback to be executed when the promise is resolved.
+    #
+    # @example
+    #   Airbrake::Promise.new.then { |val| puts val }
+    #
+    # @yield [notice_id] The ID of a notice at Airbrake
+    # @yieldparam notice_id [String]
+    # @return [self]
+    def then(&block)
+      @mutex.synchronize do
+        if @value
+          yield(@value)
+          return self
+        end
+
+        @on_resolved << block
+      end
+
+      self
+    end
+
+    ##
+    # Attaches a callback to be executed when the promise is rejected.
+    #
+    # @example
+    #   Airbrake::Promise.new.rescue { |reason| raise reason['error'] }
+    #
+    # @yield [error] The error message from the API
+    # @yieldparam error [Hash<String,String>]
+    # @return [self]
+    def rescue(&block)
+      @mutex.synchronize do
+        if @reason
+          yield(@reason)
+          return self
+        end
+
+        @on_rejected << block
+      end
+
+      self
+    end
+
+    ##
+    # Resolves the promise.
+    #
+    # @example
+    #   Airbrake::Promise.new.resolve(123)
+    #
+    # @param notice_id [String]
+    # @return [self]
+    def resolve(notice_id)
+      @mutex.synchronize do
```

Comment:
FYI does not need the synchronize since it only reads ... but does not really matter